pkg/util/validation/limits.go

```diff
@@ -176,6 +178,7 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 	f.IntVar(&l.MaxGlobalMetadataPerMetric, MaxMetadataPerMetricFlag, 0, "The maximum number of metadata per metric, across the cluster. 0 to disable.")
 	f.IntVar(&l.MaxGlobalExemplarsPerUser, "ingester.max-global-exemplars-per-user", 0, "The maximum number of exemplars in memory, across the cluster. 0 to disable exemplars ingestion.")
 	f.Var(&l.ActiveSeriesCustomTrackersConfig, "ingester.active-series-custom-trackers", "Additional active series metrics, matching the provided matchers. Matchers should be in form <name>:<matcher>, like 'foobar:{foo=\"bar\"}'. Multiple matchers can be provided either providing the flag multiple times or providing multiple semicolon-separated values to a single flag.")
+	f.DurationVar(&l.OOOAllowance, "ingester.ooo-allowance", 0*time.Second, "Allow samples to be this old for out-of-order. Supported units: h, m, s. If the value is non-zero, then the ingester will be enabled to query overlapping blocks.")
```

Comment:
No need to explain allowed units here, documentation generator already does that.

Suggested help string:

"Non-zero value Enables out-of-order support for most recent samples in this time window. Higher values need more memory. When enabled ingester will be enabled to query overlapping blocks."

(Can we explain how much memory is required for ooo samples?)

Reply:
> Can we explain how much memory is required for ooo samples

We can say that it has some impact on memory depending on frequency of ooo samples per series. We don't have any % range in hand to mention yet.

Reply:
Same question as above, what does old mean in this context? does it mean:

a) older than wall clock time
b) older than the newest ingested sample of this tenant
c) older than the newest ingested sample of this ingester

Reply:
There are 2 cases, it is either of
1) Older than the newest ingester sample for that particular series of this tenant if it exists in the Head block
2) If that series does not exist in the Head block, then it is older than the newest ingested sample of this tenant.

I think this is too detailed for config file help text? I think we will have docs for this, so can we put those details there or do you prefer it to be here?

Reply:
> "Non-zero value Enables out-of-order support for most recent samples in this time window. Higher values need more memory. When enabled ingester will be enabled to query overlapping blocks."

I will skip the `Higher values need more memory` part because it is not tied to the allowance but the nature of data. (In our beta test, 30m and 2h allowance had similar memory usage because most ooo samples were under 10 mins).

I am thinking of this:
"Non-zero value enables out-of-order support for most recent samples in this time window. Ingester will need more memory that is a factor of rate of out of order sample being ingested and number of series getting out of order samples. When enabled ingester will be enabled to query overlapping blocks."

Reply:
> "Non-zero value enables out-of-order support for most recent samples in this time window. Ingester will need more memory that is a factor of rate of out of order sample being ingested and number of series getting out of order samples. When enabled ingester will be enabled to query overlapping blocks."

Small modifications:

"Non-zero value enables out-of-order support for ~most recent~ samples in this time window. Ingester will need more memory ~that is a factor~ based on rate of out-of-order sample being ingested and number of series getting out-of-order samples. When enabled ingester will be enabled to query overlapping blocks."

WDYT?